### ISSUE TITLE
"You've found a bug in the Crystal compiler" in crystal docs because of .git directory

### ISSUE THREAD

**Issue by stefanwille** [manastech/crystal#659]
Hello,

I get this crash for `crystal docs`:

```
$ crystal docs
Nil assertion failed
*raise<String>:NoReturn +70 [4351224710]
*Nil#not_nil!<Nil>:NoReturn +13 [4351295917]
*Crystal::Command::run<Array(String)>:(Nil | CFileIO | File | Bool | Crystal::Compiler::Result | Array(Crystal::Init::View+:Class)) +14837 [4351258341]
main +15985 [4351241425]

Error: you've found a bug in the Crystal compiler. Please open an issue: https://github.com/manastech/crystal/issues
```

My source code is just this class in `src/redis.cr`:

```crystal
class Redis
end
```

However, I also have a `.git` directory with ~85 commits. The crash only occurs with this `.git` directory. If I remove it, the problem goes away. 

The whole project directory with just the source code and the `.git` directory is available here: <http://uploads.stefanwille.com/crystal/minimal-project-directory.tar.gz> (2.9 MB).
This should make it easy to reproduce.

My Crystal version is: 

```
$ crystal --version
Crystal 0.7.1 [86d387c] (Fri May  1 01:04:41 UTC 2015)
```

I am on MacOS X 10.10.3.

**Comment by asterite** [comment]
Thanks for finding this!

It turns out that [this code](https://github.com/manastech/crystal/blob/master/src/compiler/crystal/tools/doc/generator.cr#L241-L253) was misbehaving. With this I mean, the compiler generated incorrect code for that. For example, try this code in Ruby and Crystal:

```crystal
[1, 2].each do |x|
  puts x
  next if 1 == 1
  break
end
```

You'll see in Crystal it incorrectly prints 1 forever. It basically happens when there's an unconditional `break` at the end of a block, but there's a conditional `next` before it. That's why we probably never tripped with it before, because writing that is not very common.

With your report I found like 3 or 4 bugs related to this (they were similar, but different) and I fixed them. Luckily this all led to a big [code simplification](https://github.com/manastech/crystal/commit/0e7e850210041e218329e5ca87a50ecdc24056fc) because as the compiler code's evolves and becomes smarter some things that we did in one way now can be done in another way :-)

I also rewrote the buggy code in another way so the next release has the fix.
